ast.go

```diff
@@ -3575,7 +3575,19 @@ type NumberLiteral struct {
 }
 
 // String returns a string representation of the literal.
-func (l *NumberLiteral) String() string { return strconv.FormatFloat(l.Val, 'f', 3, 64) }
+func (l *NumberLiteral) String() string {
+	s := strconv.FormatFloat(l.Val, 'f', -1, 64)
+	// If l.Val is a whole number, s will not have a decimal point.
+	// Parsing the resulting string will result in a different AST with an IntegerLiteral
+	// instead of a NumberLiteral. Detect and correct this situation.
+	if strings.IndexByte(s, '.') >= 0 {
```

Comment:
why not simply `strings.IndexByte(s, '.') != -1`.  It is defined to return -1 when not found and >= 0 if found.

Reply:
maybe it is better to test for `NaN`, `+/-Inf` etc. first:

```
switch s {
 case "NaN", "+Inf", "-Inf":
    // there might be more cases than these
    return s
 }

if strings.IndexByte(s, '.') == -1 {
  // couldn't find a period
  s += ".0"
}

return 0
```

Reply:
> why not simply `strings.IndexByte(s, '.') != -1`. It is defined to return -1 when not found and >= 0 if found.

The main reason is so I don't have to remember what the exact return value is when a match isn't found :smile: .

Going out into the weeds, a general rule of thumb is to prefer operations against zero over non-zero. Most architectures and/or ABIs are slightly more efficient when zeros are involved. For instance, here's the disassembly when checking against -1:
![indexbyte-ne-1](https://github.com/influxdata/influxql/assets/10836954/e41d99f8-4a74-42c9-8cbc-2920fbe9929d)

and against 0:
![indexbyte-ge0](https://github.com/influxdata/influxql/assets/10836954/db432178-7e38-4438-b037-7c61894a2133)

The difference in this case is pretty insignificant, but checking against 0 is slightly more compact code, although the execution time should be the same on modern x86-64 implementations.

Reply:
> maybe it is better to test for `NaN`, `+/-Inf` etc. first:
> 
> ```
> s := strconv.FormatFloat(l.Val, 'f', -1, 64)
> switch s {
>  case "NaN", "+Inf", "-Inf":
>     // there might be more cases than these
>     return s
>  }
> 
> if strings.IndexByte(s, '.') == -1 {
>   // couldn't find a period
>   s += ".0"
> }
> 
> return s
> ```

I would expect NaN, +Inf, and -Inf to be pretty rare for any FP code. Since these are values are the reduced literals in an InfluxQL statement, they should be almost nonexistent in real world InfluxQL statements. I would expect the fractional case to be the most common, followed by whole numbers, with the special case FP values far behind, so waiting as long as possible to compare against those should yield better performance.

Reply:
But checking against -1 is testing against a known, documented value.  testing against >= 0 is a micro-optimization and premature optimization at best.

But -- you seem to have this handled.  I'll bow out of this PR.

Reply:
@ayang64 One really nice thing to note is that the new code is actually _faster_ than the old code. While this seems counterintuitive, `strconf.FormatFloat` will use a [faster algorithm](https://dl.acm.org/doi/10.1145/3296979.3192369) to do the conversion when -1 is passed for the precision. So even with the extra checks and work the new code is faster.

Reply:
I don't have any problem with using `strconf.FormatFloat`.  I'm just suggesting that the code path is kind of awkward in that you're potentially scanning an invalid string for periods.

Again, I'm not sure why you feel the need to implement microoptimizations vs making the code clear in a non-performance critical piece of code.

Reply:
> But checking against -1 is testing against a known, documented value. testing against >= 0 is a micro-optimization and premature optimization at best.
> 
> But -- you seem to have this handled. I'll bow out of this PR.

Feedback is always welcome. Thank you!

It's also documented to return >= 0 when the value is found, so either way seems fine to me.

While the zero wasn't a micro-optimization (it's a habit), I will admit to using `strings.IndexByte` as a micro-optimization. It benchmarked slightly faster than `strings.IndexRune` and an explicit for loop :sweat_smile: .

Reply:
Ya -- indexbyte should work just fine since the '.' isn't multi-byte.  no need to break it down into runes.  That's actually reasonable.